### PR TITLE
Implement whenSourceIsNull and fallbacks for enum mappings

### LINKED ISSUE #55: Support import aliases
I sometimes have identically named models in an api and domain package. Therefore I need to alias the import of the api package. This unfortunately generates the model names without the import alias resulting in invalid mapping code.

```dart
import 'package:api/api.dart' as api;
import 'package:auto_mappr_annotation/auto_mappr_annotation.dart';
import 'package:domain/domain.dart';

part 'test_mapper.g.dart';

@AutoMappr([
  MapType<Test, api.Test>(),
  MapType<api.Test, Test>(),
])
class Mappr extends $Mappr {}
```

### LINKED ISSUE #56: Enum fallback value
I'd like to define a fallback enum value for "unknown" values.

### PR DESCRIPTION
**Pull request by DevNico**
Close #56 

This PR also fixes a small issue with using top-level functions for whenSourceIsNull

### CODE REVIEW

**Comment by tenhobi** [comment]
Thanks a lot! 🎉 Will do technical review later, but also update readme https://github.com/netglade/auto_mappr#enum-mapping and add changes to changelog to `## Unreleased` section on top. 🙏

**Comment by DevNico** [comment]
@tenhobi Done 👍🏻

File: packages/auto_mappr/lib/src/extensions/executable_element_extension.dart [hunk 1]
```diff
@@ -1,5 +1,12 @@
 import 'package:analyzer/dart/element/element.dart';
 
 extension ExecutableElementExtension on ExecutableElement {
-  String get referCallString => isStatic ? '${enclosingElement.displayName}.$displayName' : displayName;
+  String get referCallString => isStatic &&
+          // A top-level function is static but the enclosing element has no
+          // display name. This could result in the expression being
+          // `.someFunction()` which is invalid. Therefore, we only add the
+          // enclosing element's display name if it is not empty.
```
Comment (tenhobi) [inline packages/auto_mappr/lib/src/extensions/executable_element_extension.dart:8]: Nice catch, we probably use only static methods from Mappr so we didn't catch this. Would be adding a test for this also a good idea?

Nit: could you format the text so it is one sentence per line and if the sentence is too long you break it to meaningful parts? Wanted to post a link to a guide I try to follow but damn I cannot find it. 😅

File: packages/auto_mappr/test/integration/fixture/enum_mapping.dart [hunk 2]
```diff
@@ -34,8 +38,15 @@ enum EnhancedTarget {
...
+
 @AutoMappr([
   MapType<Person, User>(),
-  MapType<EnhancedSource, EnhancedTarget>(),
+  MapType<RemotePerson, LocalPerson>(
+    whenSourceIsNull: _localRemoteUnknownDefault,
+  ),
+  MapType<EnhancedSource, EnhancedTarget>(
+    whenSourceIsNull: EnhancedTarget.employee,
+  ),
```
Comment (tenhobi) [inline packages/auto_mappr/test/integration/fixture/enum_mapping.dart:50]: There should probably also be a test maptype without `whenSourceIsNull`.

File: packages/auto_mappr/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,8 @@
+## Unreleased
+- Add when `whenSourceIsNull` support to enum mapping. (By @DevNico) [#59](https://github.com/netglade/auto_mappr/pull/59)
+- Add enum mapping fallback support for missing fields in the Target using `whenSourceIsNull`. (By @DevNico) [#59](https://github.com/netglade/auto_mappr/pull/59)
+- Allow top-level functions for `whenSourceIsNull`. (By @DevNico) [#59](https://github.com/netglade/auto_mappr/pull/59)
```
Comment (tenhobi) [inline packages/auto_mappr/CHANGELOG.md:4]: Should probably start with "Fix".

**Review by tenhobi — CHANGES_REQUESTED** [review]
Looks good, just some minor things. 🙌

File: packages/auto_mappr/lib/src/builder/auto_mappr_builder.dart [hunk 4]
```diff
@@ -17,6 +17,7 @@ class AutoMapprBuilder {
     'non_constant_identifier_names',
     'require_trailing_commas',
     'unnecessary_raw_strings',
+    'unnecessary_lambdas',
```
[the remark this answers is not in the document]
Comment (DevNico) [reply packages/auto_mappr/lib/src/builder/auto_mappr_builder.dart:20]: This is due to the whenIsNullExpression being formatted as a function call e.g. `abc()` way before we use it in the enums firstWhere orElse part. It results in generating `() => abc()`. I felt like the problems I could introduce by trying to somehow fix this issue outweighed the drawbacks of having the unnecessary lambda.

File: packages/auto_mappr/lib/src/extensions/executable_element_extension.dart (hunk 1 above)
Comment (tenhobi) [reply packages/auto_mappr/lib/src/extensions/executable_element_extension.dart:8]: Ha, https://sembr.org !

File: packages/auto_mappr/lib/src/builder/auto_mappr_builder.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (tenhobi) [reply packages/auto_mappr/lib/src/builder/auto_mappr_builder.dart:20]: Hmm, I have fixed that for other defaults and customs I think. But it seems that because orElse takes a function, and whenSourceIsNull is either a function or a const, we would have to switch between doing a lambda returning const vs calling tear off, right?

File: packages/auto_mappr/lib/src/extensions/executable_element_extension.dart (hunk 1 above)
Comment (DevNico) [reply packages/auto_mappr/lib/src/extensions/executable_element_extension.dart]: Added some tests for whenSourceIsNull which explicitly test static / top-level functions and static values.

File: packages/auto_mappr/lib/src/builder/auto_mappr_builder.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (DevNico) [reply packages/auto_mappr/lib/src/builder/auto_mappr_builder.dart:20]: Exactly I looked at the other implementations but this one is a little special :D

[the remark this answers is not in the document]
Comment (tenhobi) [reply packages/auto_mappr/lib/src/builder/auto_mappr_builder.dart:20]: We will keep it this way as knowing if that is a const value or a function is a bit hard. 👍

File: packages/auto_mappr/lib/src/extensions/executable_element_extension.dart (hunk 1 above)
Comment (tenhobi) [reply packages/auto_mappr/lib/src/extensions/executable_element_extension.dart]: We already have `default_target` and `default_field` tests, so I would instead (then creating another test that tests basically the same) just make sure that both of these files have at least one top-level function aka moving one function from Mappr class to a top-level.

File: packages/auto_mappr/lib/src/extensions/executable_element_extension.dart [hunk 5]
```diff
@@ -1,5 +1,12 @@
 import 'package:analyzer/dart/element/element.dart';
 
 extension ExecutableElementExtension on ExecutableElement {
-  String get referCallString => isStatic ? '${enclosingElement.displayName}.$displayName' : displayName;
+          // For top-level functions `isStatic` is true
+          // but the enclosing element has no display name.
+          // This could result in the expression being `.someFunction()` which is invalid.
+          // Therefore, we only add the enclosing element's display name if it is not empty.
```
[the remark this answers is not in the document]
Comment (tenhobi) [reply packages/auto_mappr/lib/src/extensions/executable_element_extension.dart]: Looks better. Just a last nit: for some reason, there are like 10 leading spaces before comments instead of like 2, and it might be rather a doc comment?

File: packages/auto_mappr/README.md [hunk 6]
```diff
@@ -392,8 +392,26 @@ class Mappr extends $Mappr {}
 AutoMappr also supports mapping of enums.
 You register them as usual with `MapType<SourceEnum, TargetEnum`>
 and AutoMappr will convert enum options based on name.
-Note that a target enum must be superset of a source enum
--- otherwise this will output a generation error.
+
+The target enum can either be a superset of the source enum or you can also
+define `whenSourceIsNull` which will be used for unknown enum values. If the
+target enum is not a superset of the source enum the generator will throw an
+error.
```
Comment (tenhobi) [inline packages/auto_mappr/README.md:399]: sembr here as well please

File: packages/auto_mappr/lib/src/extensions/executable_element_extension.dart (hunk 1 above)
Comment (DevNico) [reply packages/auto_mappr/lib/src/extensions/executable_element_extension.dart]: That would cover the edge case but the test wouldn't be there for it so refactoring down the line might remove the test case since it doesn't exit explicitly for that use case. That's just my (very minor) concern with going that way

File: packages/auto_mappr/lib/src/extensions/executable_element_extension.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (DevNico) [reply packages/auto_mappr/lib/src/extensions/executable_element_extension.dart]: Already fixed the formatting in https://github.com/netglade/auto_mappr/pull/59/commits/badcdaca80e38d4135b2a71077bbbc094645fc63 and since the comment isn't relevant for consumers of the function it's intentionally a double slash one just for someone who might edit the code later.

[the remark this answers is not in the document]
Comment (tenhobi) [reply packages/auto_mappr/lib/src/extensions/executable_element_extension.dart]: Seems all right.

**Review by tenhobi — APPROVED** [review]
jiffy

**Review by petrnymsa — APPROVED** [review]
LGTM.
